torchdata/datapipes/iter/transform/callable.py

```diff
@@ -414,3 +418,138 @@ def __len__(self) -> int:
+
+
+class _BatchAsyncMapperIterDataPipe(IterDataPipe):
+    datapipe: IterDataPipe
+    async_fn: Callable
+
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        async_fn: Callable,
+        input_col=None,
+        output_col=None,
+        max_concurrency: int = 32,
+    ):
+        self.source_datapipe = source_datapipe
+        if not inspect.iscoroutinefunction(async_fn):
+            raise ValueError(f"Expected a corotine function with an async def syntax, but got a {type(async_fn)}")
+        self.async_fn = async_fn  # type: ignore[assignment]
+        if input_col is None and output_col is not None:
+            raise ValueError("`output_col` must be None when `input_col` is None.")
+        self.input_col = input_col
+        if isinstance(output_col, (list, tuple)):
+            if len(output_col) > 1:
+                raise ValueError("`output_col` must be a single-element list or tuple")
+            output_col = output_col[0]
+        self.output_col = output_col
+        self.max_concurrency = max_concurrency
+
+    def __iter__(self):
+        for batch in self.source_datapipe:
+            new_batch = asyncio.run(self.processbatch(batch))
+            yield new_batch
+
+    async def processbatch(self, batch):
+        sem = asyncio.Semaphore(self.max_concurrency)
+
+        async def controlled_async_fn(async_fn, *data):
+            async with sem:
+                return await async_fn(*data)
+
+        coroutines = []
+        if self.input_col is None:
+            for data in batch:
+                coroutines.append(controlled_async_fn(self.async_fn, data))
+            results = await asyncio.gather(*coroutines)
+            return results
+
+        for data in batch:
+            if isinstance(self.input_col, (list, tuple)):
+                args = tuple(data[col] for col in self.input_col)
+                coroutines.append(controlled_async_fn(self.async_fn, *args))
+            else:
+                coroutines.append(controlled_async_fn(self.async_fn, data[self.input_col]))
+        results = await asyncio.gather(*coroutines)
+
+        new_batch = []
+        for data, res in zip(batch, results):
+            t_flag = isinstance(data, tuple)
+            if t_flag:
+                data = list(data)
+
+            if self.output_col is None:
+                if isinstance(self.input_col, (list, tuple)):
+                    data[self.input_col[0]] = res
+                    for idx in sorted(self.input_col[1:], reverse=True):
+                        del data[idx]
+                else:
+                    data[self.input_col] = res
+            elif self.output_col == -1:
+                data.append(res)
+            else:
+                data[self.output_col] = res
+
+            if t_flag:
+                data = tuple(data)
+
+            new_batch.append(data)
+        return new_batch
+
+
+@functional_datapipe("async_map_batches")
```

Comment:
map_batches, lol: https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.map_batches.html#ray.data.Dataset.map_batches

Reply:
Yeah, any naming suggestion is welcome